File: bench/tokenizer_bench.zig
//! tokenizer_bench.zig
//!
//! Measures the throughput of simd.zig::findJsonStructure.
//! This is the apples-to-apples comparison with simd.cpp (simdjson stage1).
//!
//! What is being measured
//! ──────────────────────
//!   Input : raw bytes of a NDJSON file (already mmap'd — zero-copy into mem)
//!   Output: count of structural tokens found  { } [ ] " : ,
//!   Metric: GB/s
//!
//! The file is mapped once OUTSIDE the timing loop so disk I/O is excluded.
//! Each iteration scans the entire in-memory buffer from scratch.
//!
//! Usage: zig build bench-tokenizer -- bench/bench_data.ndjson [iterations]

const std = @import("std");
const simd = @import("simd"); // imported via build.zig module alias

const MAX_TOKENS = 64 * 1024 * 1024; // 64 M tokens — enough for 1 M records × ~60 tokens each

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.c_allocator);
    defer arena.deinit();
    const alloc = arena.allocator();

    // ── Parse args ────────────────────────────────────────────────────────────
    const args = try std.process.argsAlloc(alloc);
    if (args.len < 2) {
        std.debug.print("Usage: {s} <file.ndjson> [iterations]\n", .{args[0]});
        return error.MissingArgument;
    }
    const filename = args[1];
    const iters: usize = if (args.len >= 3) try std.fmt.parseInt(usize, args[2], 10) else 5;

    // ── Open & mmap file ──────────────────────────────────────────────────────
    const file = try std.fs.cwd().openFile(filename, .{});
    defer file.close();
    const file_size = (try file.stat()).size;

    // Use mmap for zero-copy I/O (same as the application path)
    const data = try std.posix.mmap(
        null,
        file_size,
        std.posix.PROT.READ,
        .{ .TYPE = .PRIVATE },
        file.handle,
        0,
    );
    defer std.posix.munmap(data);

    std.debug.print("File loaded : {s}\n", .{filename});
    std.debug.print("File size   : {d:.3} GB  ({d} bytes)\n", .{ @as(f64, @floatFromInt(file_size)) / 1e9, file_size });

    // ── Allocate token buffer (reused across iterations) ──────────────────────
    const tokens = try alloc.alloc(simd.Token, MAX_TOKENS);
    std.debug.print("Token buf   : {d:.1} MB  (max {d} tokens)\n\n", .{ @as(f64, @floatFromInt(tokens.len * @sizeOf(simd.Token))) / 1e6, tokens.len });

    // ── Warm-up (1 pass, not timed) ───────────────────────────────────────────
    _ = simd.findJsonStructure(data, tokens);

    // ── Timed iterations ──────────────────────────────────────────────────────
    std.debug.print("Running {d} timed iteration(s)...\n", .{iters});

    var best_secs: f64 = std.math.inf(f64);
    var total_secs: f64 = 0;
    var total_tokens: usize = 0;
    var run_times = try alloc.alloc(f64, iters);

    for (0..iters) |iter| {
        @memset(tokens, std.mem.zeroes(simd.Token));

        const t0 = std.time.nanoTimestamp();
        const n = simd.findJsonStructure(data, tokens);
        const elapsed = @as(f64, @floatFromInt(std.time.nanoTimestamp() - t0)) / 1e9;

        run_times[iter] = elapsed;
        total_secs += elapsed;
        total_tokens += n;

        if (elapsed < best_secs) best_secs = elapsed;

        std.debug.print("  iter {d:2}: {d:.4}s  tokens={d}  {d:.2} GB/s\n", .{ iter + 1, elapsed, n, @as(f64, @floatFromInt(file_size)) / 1e9 / elapsed });
    }

    const avg_secs = total_secs / @as(f64, @floatFromInt(iters));
    const tokens_per_iter = total_tokens / iters;

    const gb = @as(f64, @floatFromInt(file_size)) / 1e9;
    std.debug.print(
        \\
        \\── Summary ──────────────────────────────────────────────────────────
        \\  tokens/iter   : {d}
        \\  best run      : {d:.4}s  →  {d:.2} GB/s
        \\  avg  run      : {d:.4}s  →  {d:.2} GB/s
        \\
    , .{
        tokens_per_iter,
        best_secs,
        gb / best_secs,
        avg_secs,
        gb / avg_secs,
    });

    // Machine-readable line for the comparison script (stdout)
    var out_buf: [256]u8 = undefined;
    const out_line = try std.fmt.bufPrint(
        &out_buf,
        "zson_gb_per_sec={d:.2} zson_best_sec={d:.4} zson_tokens={d}\n",
        .{ gb / best_secs, best_secs, tokens_per_iter },
    );
    _ = try std.posix.write(std.posix.STDOUT_FILENO, out_line);
}

File: bench/tokenizer_bench.cpp
/**
 * tokenizer_bench.cpp
 *
 * Measures simdjson's structural character finding (stage1) throughput.
 * This is the apples-to-apples comparison against simd.zig::findJsonStructure.
 *
 * Both programs do the same task:
 *   Input : raw bytes of NDJSON data (already in memory)
 *   Output: positions of structural characters { } [ ] " : ,
 *   Metric: GB/s  (number of input bytes processed per second)
 *
 * The file is loaded ONCE outside the timing loop.
 * Each iteration re-parses the entire buffer from the same in-memory copy.
 *
 * Compile:
 *   c++ -O3 -std=c++17 -DSIMDJSON_THREADS_ENABLED=1 \
 *       -I/opt/homebrew/Cellar/simdjson/4.2.4/include \
 *       -L/opt/homebrew/Cellar/simdjson/4.2.4/lib -lsimdjson \
 *       -o bench/tokenizer_bench bench/tokenizer_bench.cpp
 *
 * Usage:
 *   ./bench/tokenizer_bench <file.ndjson> [iterations]
 */

#include <simdjson.h>
#include <cassert>
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <vector>

static double now_secs() {
    using namespace std::chrono;
    return duration<double>(high_resolution_clock::now().time_since_epoch()).count();
}

int main(int argc, char** argv) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s <file.ndjson> [iterations]\n", argv[0]);
        return 1;
    }
    const char* filename = argv[1];
    const int iters = argc >= 3 ? atoi(argv[2]) : 5;

    // ── Load file once into a padded_string ──────────────────────────────────
    simdjson::padded_string json;
    auto err = simdjson::padded_string::load(filename).get(json);
    if (err) {
        fprintf(stderr, "Load error: %s\n", simdjson::error_message(err));
        return 1;
    }
    const size_t file_size = json.size();
    fprintf(stderr, "File loaded: %.3f GB  (%zu bytes)\n",
            (double)file_size / 1e9, file_size);

    // ── Warm-up (1 iteration, not timed) ─────────────────────────────────────
    {
        simdjson::ondemand::parser parser;
        simdjson::ondemand::document_stream stream;
        if (!parser.iterate_many(json, 1 << 20).get(stream)) {
            for (auto it = stream.begin(); it != stream.end(); ++it) {
                auto doc = *it;
                (void)doc;
            }
        }
    }

    // ── Timed iterations ─────────────────────────────────────────────────────
    // We measure the full ondemand parse loop (iterate_many) — this is what
    // simdjson does under the hood: stage1 (find structural bits) + stage2
    // (lazy value resolution on demand).  For NDJSON filtering workloads,
    // stage1 dominates.  Each iteration creates a fresh parser + stream so
    // that CPU caches see the data fresh each time.
    fprintf(stderr, "Running %d timed iteration(s)...\n", iters);

    double total_time = 0.0;
    size_t total_docs = 0;
    std::vector<double> run_times;

    for (int iter = 0; iter < iters; ++iter) {
        simdjson::ondemand::parser parser;
        simdjson::ondemand::document_stream stream;
        size_t docs = 0;

        double t0 = now_secs();

        if (!parser.iterate_many(json, 1 << 20).get(stream)) {
            for (auto it = stream.begin(); it != stream.end(); ++it) {
                // Access ONE field to force simdjson to actually tokenize the
                // document (lazy parsing doesn't tokenize until accessed).
                // We use a dummy key that won't exist — simdjson still runs
                // stage1 + stage2 header parsing to find the field.
                auto doc = *it;
                double val;
                (void)(doc["__z__"].get(val)); // always MISS — but forces parse
                ++docs;
            }
        }

        double elapsed = now_secs() - t0;
        run_times.push_back(elapsed);
        total_time += elapsed;
        total_docs += docs;
    }

    // ── Report ────────────────────────────────────────────────────────────────
    double best = run_times[0];
    for (double t : run_times) if (t < best) best = t;
    double avg = total_time / iters;

    fprintf(stderr,
        "\nsimdjson v%s  ondemand/iterate_many  (stage1+stage2, lazy field)\n"
        "  file_size : %.3f GB\n"
        "  iters     : %d\n"
        "  docs/iter : %zu\n"
        "  best run  : %.4fs  →  %.2f GB/s\n"
        "  avg  run  : %.4fs  →  %.2f GB/s\n",
        SIMDJSON_VERSION,
        (double)file_size / 1e9,
        iters,
        total_docs / (size_t)iters,
        best,  (double)file_size / 1e9 / best,
        avg,   (double)file_size / 1e9 / avg);

    // Machine-readable line for the comparison script
    printf("simdjson_gb_per_sec=%.2f simdjson_best_sec=%.4f simdjson_docs=%zu\n",
           (double)file_size / 1e9 / best, best, total_docs / (size_t)iters);

    return 0;
}
